### Урок 2. Анализ датасета с помощью Pandas
#### Условие 1: 1 задача
Скачать файл в уроке

Считать данные с помощью pandas

Вывести на экран первые 5 строк



In [129]:

import pandas as pd


In [130]:
df = pd.read_csv("kc_house_data.csv", sep = ',', encoding= 'windows - 1251')
df.head(5)


id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7      1180.0              0   
1      7242     2.0           0     0  ...      7      2170.0            400   
2     10000     1.0           0     0  ...      6       770.0              0   
3      5000     1.0           0     0  ...      7      1050.0            910   
4      8080     1.0           0     0  ...      8      1680.0              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

Посмотреть на описание признаков и на их содержание


In [131]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21611 non-null  float64
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

Проведите первичный анализ данных
Изучите типы данных
Найдите количество пропущенных ячеек в данных
Посчитайте основные статистики по всем признакам и поизучайте их

In [132]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21611.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.396095     291.509045   1971.005136     84.402258   
std        1.175459    828.128162     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [133]:
df.describe(include = ['object'])

date
count             21613
unique              372
top     20140623T000000
freq                142

Пропущенных объектов нет, 372 уникальных объекта, частота появления самоготчастотного значения - 142, есть целочисленные значения, дробные

3.1 В каком диапазоне изменяются стоимости недвижимости?

In [134]:
df['price'].value_counts()

price
350000.0     172
450000.0     172
550000.0     159
500000.0     152
425000.0     150
            ... 
607010.0       1
1362500.0      1
298800.0       1
957500.0       1
402101.0       1
Name: count, Length: 4028, dtype: int64

In [135]:
df.head(0)

Empty DataFrame
Columns: [id, date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

[0 rows x 21 columns]

3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [136]:
df['sqft_living'].mean()

2079.8997362698374

3.3 Как много домов с разными этажами в данных?

In [137]:
df['floors'].nunique()

6

3.4 Насколько хорошие состояния у домов в данных?

In [138]:
df['condition'].mean()

3.4094295100171195


3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [139]:
df['yr_built'][0]

1955

In [140]:
df.tail(1)

id             date     price  bedrooms  bathrooms  \
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
21612      1020.0              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
21612 -122.299           1020        1357  

[1 rows x 21 columns]

In [141]:
df['yr_built'][21612]

2008

4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [142]:
df[df['bedrooms'] == 2 ]['price'].mean()

401372.681884058

4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?



In [143]:
df[df['price'] > 600000 ]['sqft_lot'].mean()

20442.524776214832

4.3 Как много домов коснулся ремонт?  
# ответ 914, но я не знаю, как правильно записать

In [144]:
df[~(df['yr_renovated'] == 0)]

id             date      price  bedrooms  bathrooms  \
1      6414100192  20141209T000000   538000.0         3       2.25   
35     9547205180  20140613T000000   696000.0         3       2.50   
95     1483300570  20140908T000000   905000.0         4       2.50   
103    2450000295  20141007T000000  1088000.0         3       2.50   
115    3626039325  20141121T000000   740500.0         3       3.50   
...           ...              ...        ...       ...        ...   
19622  7351200295  20150114T000000  1150000.0         3       1.75   
20057   126039256  20140904T000000   434900.0         3       2.00   
20444  4305600360  20150225T000000   500012.0         4       2.50   
20447  3319500628  20150212T000000   356999.0         3       1.50   
20962  1278000210  20150311T000000   110000.0         2       1.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
1             2570      7242     2.0           0     0  ...      7   
35            2300      3060     1.5           0     0  ...      8   
95            3300     10250     1.0           0     0  ...      7   
103           2920      8113     2.0           0     0  ...      8   
115           4380      6350     2.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
19622         1760      6788     2.0           1     4  ...      7   
20057         1520      5040     2.0           0     0  ...      7   
20444         2400      9612     1.0           0     0  ...      8   
20447         1010      1546     2.0           0     0  ...      8   
20962          828      4524     1.0           0     0  ...      6   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
1          2170.0            400      1951          1991    98125  47.7210   
35         1510.0            790      1930          2002    98115  47.6827   
95         2390.0            910      1946          1991    98040  47.5873   
103        2920.0              0      1950          2010    98004  47.5814   
115        2780.0           1600      1900          1999    98117  47.6981   
...           ...            ...       ...           ...      ...      ...   
19622      1760.0              0      1940          1960    98125  47.7336   
20057      1520.0              0      1977          2006    98177  47.7770   
20444      1230.0           1170      1962          2009    98059  47.4799   
20447      1010.0              0      1971          2014    98144  47.5998   
20962       828.0              0      1968          2007    98001  47.2655   

          long  sqft_living15  sqft_lot15  
1     -122.319           1690        7639  
35    -122.310           1590        3264  
95    -122.249           1950        6045  
103   -122.196           2370        8113  
115   -122.368           1830        6350  
...        ...            ...         ...  
19622 -122.284           1630        7588  
20057 -122.362           1860        8710  
20444 -122.127           2430        5539  
20447 -122.311           1010        1517  
20962 -122.244            828        5402  

[914 rows x 21 columns]


4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [145]:
grade_10 =df[df['grade']>= 10]['price'].mean()
grade_10

1257984.857579462

In [146]:
grade_4 =df[df['grade'] <= 4]['price'].mean()
grade_4

211395.45454545456

In [147]:
##diff = df[grade_10] - df[grade_4]
##diff 

5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?
# ОТВЕТ 41

In [148]:
df[~(df['waterfront'] == 0) & (df['bathrooms'] >= 3) & (df['sqft_basement'] > 0)]

id             date      price  bedrooms  bathrooms  \
300    3225069065  20140624T000000  3075000.0         4       5.00   
457    2122039094  20141126T000000   705000.0         3       3.00   
1164   1247600105  20141020T000000  5110800.0         5       5.25   
1261   7636800041  20140625T000000   995000.0         3       4.50   
1315   7558700030  20150413T000000  5300000.0         6       6.00   
1361   1925069082  20150511T000000  2200000.0         5       4.25   
2626   7738500731  20140815T000000  4500000.0         5       5.50   
2864   4114601570  20141118T000000  3600000.0         3       3.25   
3914   9808700762  20140611T000000  7062500.0         5       4.50   
3934   1269200229  20140723T000000  1379900.0         3       3.25   
5760    623059016  20140717T000000  1100000.0         4       3.25   
6233   2024069008  20140619T000000  2200000.0         5       4.75   
6508   4217402115  20150421T000000  3650000.0         6       4.75   
7989   9362000040  20140623T000000  3400000.0         3       4.50   
8092   1924059029  20140617T000000  4668000.0         5       6.75   
8163   1724069079  20150319T000000  1452000.0         2       3.25   
8784   3343302110  20150306T000000  1800000.0         3       3.00   
10263   625059036  20140813T000000  2700000.0         5       4.00   
11257  3761100045  20140618T000000  3000000.0         4       4.25   
11402  3426049284  20140819T000000  2300000.0         4       3.25   
11424  5119400075  20140620T000000   950000.0         3       3.25   
11446   624069035  20141209T000000  2750000.0         4       4.00   
11844  2623039082  20150218T000000   770000.0         3       3.50   
12459  4114601580  20140724T000000  1900000.0         6       4.00   
12935  3222049044  20140612T000000   835000.0         3       3.00   
13710  2923500230  20141216T000000  2600000.0         4       4.50   
13967  3126059023  20150303T000000  3395000.0         4       3.50   
14083  3625059043  20140904T000000  3300000.0         5       4.75   
14619  7936500109  20140725T000000  2230000.0         3       3.00   
14822  3623500408  20150330T000000  2600000.0         3       3.00   
15482   624069108  20140812T000000  3200000.0         4       3.25   
15668  4030100290  20141001T000000  1680000.0         5       3.50   
16817   825059178  20140923T000000  2574000.0         4       3.75   
17767  7352200450  20150115T000000  2050000.0         4       3.25   
18199  3625059152  20141230T000000  3300000.0         3       3.25   
18455  8043700300  20140608T000000  2700000.0         4       3.25   
18556  6613000930  20140902T000000  2950000.0         4       3.25   
20325   518500480  20140811T000000  3000000.0         3       3.50   
20767  8043700105  20150417T000000  2300000.0         4       4.00   
21201   518500460  20141008T000000  2230000.0         3       3.50   
21576  9253900271  20150107T000000  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
300           4550     18641     1.0           1     4  ...     10   
457           1970     20978     2.0           1     3  ...      9   
1164          8010     45517     2.0           1     4  ...     12   
1261          4380     47044     2.0           1     3  ...      9   
1315          7390     24829     2.0           1     4  ...     12   
1361          4640     22703     2.0           1     4  ...      8   
2626          6640     40014     2.0           1     4  ...     12   
2864          5020     12431     2.0           1     4  ...     10   
3914         10040     37325     2.0           1     2  ...     11   
3934          3786     38038     1.0           1     4  ...      9   
5760          3190     11774     2.0           1     4  ...      8   
6233          5990     10450     2.0           1     4  ...     11   
6508          5480     19401     1.5           1     4  ...     11   
7989          5230     17826     2.0           1     4  ...     10   
8092          9640     1

5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?
5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [157]:
df[~(df['waterfront'] == 0) | (df['view'] >= 3) | (df['yr_built'] >= 1980)]['price']

4         510000.0
5        1225000.0
6         257500.0
9         323000.0
15        650000.0
           ...    
21608     360000.0
21609     400000.0
21610     402101.0
21611     400000.0
21612     325000.0
Name: price, Length: 9804, dtype: float64

5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [160]:
df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] >= 150000)]['grade'].mean()

8.275488770399095